specification/compiler/steps/validate-model.ts

```diff
@@ -0,0 +1,594 @@
+  // Type definitions validations
+
+  function validateTypeDef (typeDef: model.TypeDefinition): void {
+    if (typesSeen.has(fqn(typeDef.name))) {
+      // Already visited
+      return
+    }
+    typesSeen.add(fqn(typeDef.name))
+
+    // Start a new context, type definitions are top-level elements
+    const oldContext = context
+    context = []
+    context.push(`${typeDef.kind} definition ${typeDef.name.namespace}:${typeDef.name.name}`)
+
+    // Validate description URL
+    if (typeDef.url != null) {
+      try {
+        new URL(typeDef.url) // eslint-disable-line no-new
+      } catch (error) {
+        modelError('Description URL is malformed')
+      }
+    }
+
+    switch (typeDef.kind) {
+      case 'request':
+        validateRequest(typeDef)
+        break
+
+      case 'interface':
+        validateInterface(typeDef)
+        break
+
+      case 'enum':
+        validateEnum(typeDef)
+        break
+
+      case 'type_alias':
+        validateTypeAlias(typeDef)
+        break
+
+      case 'union':
+        validateUnion(typeDef)
+        break
+    }
+
+    // Restore previous context
+    context = oldContext
+  }
+
+  function validateRequest (typeDef: model.Request): void {
+    const openGenerics = openGenericSet(typeDef)
+
+    validateInherits(typeDef.inherits, openGenerics)
+    validateImplements(typeDef.implements, openGenerics)
+    validateBehaviors(typeDef, openGenerics)
+
+    // Note: we validate identifier/name uniqueness independently in the path, query and body as there are some
+    // valid overlaps, with some parameters that can also be represented as body properties.
+    // Client generators will have to take care of this.
+
+    context.push('path')
+    validateProperties(typeDef.path, openGenerics)
+    context.pop()
+
+    context.push('query')
+    validateProperties(typeDef.query, openGenerics)
+    context.pop()
+
+    if (typeDef.body != null) {
+      context.push('body')
+      switch (typeDef.body.kind) {
+        case 'properties':
+          validateProperties(typeDef.body.properties, openGenerics)
+          break
+        case 'value':
+          validateValueOf(typeDef.body.value, openGenerics)
+          break
+      }
+      context.pop()
+    }
+  }
+
+  function validateInterface (typeDef: model.Interface): void {
+    const openGenerics = openGenericSet(typeDef)
+
+    validateImplements(typeDef.implements, openGenerics)
+    validateInherits(typeDef.inherits, openGenerics)
+    validateBehaviors(typeDef, openGenerics)
+    validateProperties(typeDef.properties, openGenerics)
+  }
+
+  function validateEnum (typeDef: model.Enum): void {
+    const allIdentifiers = new Set<string>()
+    const allNames = new Set<string>()
+
+    for (const item of typeDef.members) {
+      // Identifier must be unique among all items (case insensitive)
+      const identifier = (item.identifier ?? item.name).toLowerCase()
```

Comment:
I fear there are some cases in the type validator that require different casing, as the Elasticsearch API can accept multiple variations of the type.